File: tools/smxtools/smxdasm/Rtti.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace smxdasm
{
    public class SmxRttiListTable : SmxSection
    {
        protected uint header_size_;
        protected uint row_size_;
        protected uint row_count_;

        public SmxRttiListTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header)
        {
            var reader = file.SectionReader(header);
        }

        protected void init(BinaryReader reader) {
            header_size_ = reader.ReadUInt32();
            row_size_ = reader.ReadUInt32();
            row_count_ = reader.ReadUInt32();
        }
    }

    public static class cb
    {
        public const byte kBool = 0x01;
        public const byte kInt32 = 0x06;
        public const byte kFloat32 = 0x0c;
        public const byte kChar8 = 0x0e;
        public const byte kAny = 0x10;
        public const byte kTopFunction = 0x11;

        public const byte kFixedArray = 0x30;
        public const byte kArray = 0x31;
        public const byte kFunction = 0x32;

        public const byte kEnum = 0x42;
        public const byte kTypedef = 0x43;
        public const byte kTypeset = 0x44;
        public const byte kStruct = 0x45;
        public const byte kEnumStruct = 0x46;

        public const byte kVoid = 0x70;
        public const byte kVariadic = 0x71;
        public const byte kByRef = 0x72;
        public const byte kConst = 0x73;

        public const byte kTypeId_Inline = 0x0;
        public const byte kTypeId_Complex = 0x1;

        public static int DecodeUint32(byte[] bytes, ref int offset) {
            uint value = 0;
            int shift = 0;
            for (; ;) {
                byte b = bytes[offset++];
                value |= (uint)(b & 0x7f) << shift;
                if ((b & 0x80) == 0)
                    break;
                shift += 7;
            }
            return (int)value;
        }
    }

    public class SmxRttiData : SmxSection
    {
        private SmxFile smx_file_;
        private byte[] bytes_;

        public SmxRttiData(SmxFile file, FileHeader file_header, SectionEntry header)
            : base(file_header, header)
        {
            smx_file_ = file;
            var reader = file_header.SectionReader(header);
            bytes_ = reader.ReadBytes(header.Size);
        }

        public string TypeFromTypeId(int type_id)
        {
            int kind = (type_id) & 0xf;
            int payload = ((type_id) >> 4) & 0xfffffff;

            if (kind == cb.kTypeId_Inline)
            {
                byte[] temp = new byte[4];
                temp[0] = (byte)(payload & 0xff);
                temp[1] = (byte)((payload >> 8) & 0xff);
                temp[2] = (byte)((payload >> 16) & 0xff);
                temp[3] = (byte)((payload >> 24) & 0xff);

                TypeBuilder tb = new TypeBuilder(smx_file_, temp, 0);
                return tb.DecodeNew();
            }
            if (kind != cb.kTypeId_Complex)
                throw new Exception("Unknown type_id kind: " + kind);
            return BuildTypename(ref payload);
        }

        public string FunctionTypeFromOffset(int offset)
        {
            TypeBuilder b = new TypeBuilder(smx_file_, bytes_, offset);
            return b.DecodeFunction();
        }

        public string[] TypesetTypesFromOffset(int offset)
        {
            int count = cb.DecodeUint32(bytes_, ref offset);
            string[] types = new string[count];

            TypeBuilder tb = new TypeBuilder(smx_file_, bytes_, offset);
            for (int i = 0; i < count; i++)
                types[i] = tb.DecodeNew();
            return types;
        }

        private string BuildTypename(ref int offset)
        {
            TypeBuilder b = new TypeBuilder(smx_file_, bytes_, offset);
            string text = b.DecodeNew();
            offset = b.offset;
            return text;
        }

        private class TypeBuilder
        {
            private SmxFile file_;
            private byte[] bytes_;
            private int offset_;
            private bool is_const_ = false;

            public TypeBuilder(SmxFile file, byte[] bytes, int offset)
            {
                file_ = file;
                bytes_ = bytes;
                offset_ = offset;
            }

            // Decode a type, but reset the |is_const| indicator for non-
            // dependent type.
            public string DecodeNew()
            {
                bool was_const = is_const_;
                is_const_ = false;

                string result = Decode();
                if (is_const_)
                    result = "const " + result;

                is_const_ = was_const;
                return result;
            }

            private string Decode()
            {
                is_const_ |= match(cb.kConst);
                byte b = bytes_[offset_++];
                switch (b)
                {
                    case cb.kBool: return "bool";
                    case cb.kInt32: return "int";
                    case cb.kFloat32: return "float";
                    case cb.kChar8: return "char";
                    case cb.kAny: return "any";
                    case cb.kTopFunction: return "Function";
                    case cb.kFixedArray:
                    {
                        var index = cb.DecodeUint32(bytes_, ref offset_);
                        string inner = Decode();
                        return inner + "[" + index + "]";
                    }
                    case cb.kArray:
                    {
                        string inner = Decode();
                        return inner + "[]";
                    }
                    case cb.kEnum:
                    {
                        var index = cb.DecodeUint32(bytes_, ref offset_);
                        return file_.RttiEnums.Enums[index];
                    }
                    case cb.kTypedef:
                    {
                        var index = cb.DecodeUint32(bytes_, ref offset_);
                        return file_.RttiTypedefs.Typedefs[index].name;
                    }
                    case cb.kTypeset:
                    {
                        var index = cb.DecodeUint32(bytes_, ref offset_);
                        return file_.RttiTypesets.Typesets[index].name;
                    }
                    case cb.kStruct:
                    {
                        var index = cb.DecodeUint32(bytes_, ref offset_);
                        return file_.RttiClassDefs.Defs[index].name;
                    }
                    case cb.kFunction:
                        return DecodeFunction();
                    case cb.kEnumStruct:
                    {
                        var index = cb.DecodeUint32(bytes_, ref offset_);
                        return file_.RttiEnumStructs.Entries[index].name;
                    }
                }
                throw new Exception("unknown type code: " + b);
            }

            public string DecodeFunction()
            {
                uint argc = (uint)bytes_[offset_++];

                bool variadic = false;
                if (bytes_[offset] == cb.kVariadic) {
                    variadic = true;
                    offset_++;
                }

                string return_type;
                if (bytes_[offset] == cb.kVoid) {
                    return_type = "void";
                    offset_++;
                } else {
                    return_type = DecodeNew();
                }

                string[] argv = new string[argc];
                for (uint i = 0; i < argc; i++)
                {
                    bool is_byref = match(cb.kByRef);
                    string text = DecodeNew();
                    if (is_byref)
                        text += "&";
                    argv[i] = text;
                }

                string signature =
                    "function " +
                    return_type + " (" +
                    String.Join(", ", argv);
                if (variadic)
                    signature += "...";
                signature += ")";
                return signature;
            }

            private bool match(byte b)
            {
                if (bytes_[offset_] != b)
                    return false;
                offset_++;
                return true;
            }

            public int offset { get { return offset_; } }
        }
    }

    public class SmxRttiEnumTable : SmxRttiListTable
    {
        public SmxRttiEnumTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Enums = new string[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                int index = reader.ReadInt32();
                Enums[i] = names.StringAt(index);
                // reserved0-2.
                reader.ReadInt32();
                reader.ReadInt32();
                reader.ReadInt32();
            }
        }

        public string[] Enums { get; }
    }

    public struct RttiMethod
    {
        public string name;
        public int pcode_start;
        public int pcode_end;
        public int signature;
    }

    public class SmxRttiMethodTable : SmxRttiListTable
    {
        public SmxRttiMethodTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Methods = new RttiMethod[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                int index = reader.ReadInt32();
                Methods[i].name = names.StringAt(index);
                Methods[i].pcode_start = reader.ReadInt32();
                Methods[i].pcode_end = reader.ReadInt32();
                Methods[i].signature = reader.ReadInt32();
            }
        }

        public RttiMethod[] Methods { get; }
    }

    public struct RttiNative
    {
        public string name;
        public int signature;
    }

    public class SmxRttiNativeTable : SmxRttiListTable
    {
        public SmxRttiNativeTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Natives = new RttiNative[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                int index = reader.ReadInt32();
                Natives[i].name = names.StringAt(index);
                Natives[i].signature = reader.ReadInt32();
            }
        }

        public RttiNative[] Natives { get; }
    }

    public struct RttiTypedef
    {
        public string name;
        public int type_id;
    }
 
    public class SmxRttiTypedefTable : SmxRttiListTable
    {
        public SmxRttiTypedefTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Typedefs = new RttiTypedef[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                int index = reader.ReadInt32();
                Typedefs[i].name = names.StringAt(index);
                Typedefs[i].type_id = reader.ReadInt32();
            }
        }

        public RttiTypedef[] Typedefs { get; }
    }

    public struct RttiTypeset
    {
        public string name;
        public int signature;
    }
 
    public class SmxRttiTypesetTable : SmxRttiListTable
    {
        public SmxRttiTypesetTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Typesets = new RttiTypeset[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                int index = reader.ReadInt32();
                Typesets[i].name = names.StringAt(index);
                Typesets[i].signature = reader.ReadInt32();
            }
        }

        public RttiTypeset[] Typesets { get; }
    }

    public class RttiEnumStruct
    {
        public int name_offset;
        public int first_field;
        public int size;
        public string name;
    }

    public class SmxRttiEnumStructTable : SmxRttiListTable
    {
        public SmxRttiEnumStructTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Entries = new RttiEnumStruct[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                Entries[i] = new RttiEnumStruct();
                Entries[i].name_offset = reader.ReadInt32();
                Entries[i].first_field = reader.ReadInt32();
                Entries[i].size = reader.ReadInt32();
                Entries[i].name = names.StringAt(Entries[i].name_offset);
            }
        }

        public RttiEnumStruct[] Entries { get; }
    }

    public class RttiEnumStructField
    {
        public int name_offset;
        public int type_id;
        public int offset;
        public string name;
    }

    public class SmxRttiEnumStructFieldTable : SmxRttiListTable
    {
        public SmxRttiEnumStructFieldTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Entries = new RttiEnumStructField[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                Entries[i] = new RttiEnumStructField();
                Entries[i].name_offset = reader.ReadInt32();
                Entries[i].type_id = reader.ReadInt32();
                Entries[i].offset = reader.ReadInt32();
                Entries[i].name = names.StringAt(Entries[i].name_offset);
            }
        }

        public RttiEnumStructField[] Entries { get; }
    }

    public class RttiClassDef
    {
        public int flags;
        public int name_offset;
        public int first_field;
        public string name;
    }

    public class SmxRttiClassDefTable : SmxRttiListTable
    {
        public SmxRttiClassDefTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Defs = new RttiClassDef[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                Defs[i] = new RttiClassDef();
                Defs[i].flags = reader.ReadInt32();
                Defs[i].name_offset = reader.ReadInt32();
                Defs[i].first_field = reader.ReadInt32();
                Defs[i].name = names.StringAt(Defs[i].name_offset);
                // reserved0-3
                reader.ReadInt32();
                reader.ReadInt32();
                reader.ReadInt32();
                reader.ReadInt32();
            }
        }

        public RttiClassDef[] Defs { get; }
    }

    public class RttiField
    {
        public short flags;
        public int name_offset;
        public int type_id;
        public string name;
    }

    public class SmxRttiFieldTable : SmxRttiListTable
    {
        public SmxRttiFieldTable(FileHeader file, SectionEntry header, SmxNameTable names)
            : base(file, header, names)
        {
            var reader = file.SectionReader(header);
            base.init(reader);

            Fields = new RttiField[row_count_];
            for (uint i = 0; i < row_count_; i++) {
                Fields[i] = new RttiField();
                Fields[i].flags = reader.ReadInt16();
                Fields[i].name_offset = reader.ReadInt32();
                Fields[i].type_id = reader.ReadInt32();
                Fields[i].name = names.StringAt(Fields[i].name_offset);
            }
        }

        public RttiField[] Fields { get; }
    }
}

File: vm/rtti.cpp
// vim: set sts=2 ts=8 sw=2 tw=99 et:
//
// Copyright (C) 2004-2021 AlliedModers LLC
//
// This file is part of SourcePawn. SourcePawn is licensed under the GNU
// General Public License, version 3.0 (GPL). If a copy of the GPL was not
// provided with this file, you can obtain it here:
//   http://www.gnu.org/licenses/gpl.html
//
#include "rtti.h"
#include <smx/smx-typeinfo.h>

using namespace ke;
using namespace sp::debug;

RttiData::RttiData()
 : rtti_data_(nullptr),
   rtti_data_size_(0)
{
}

RttiData::RttiData(const uint8_t* blob, uint32_t size)
 : rtti_data_(blob),
   rtti_data_size_(size)
{
}

const Rtti*
RttiData::typeFromTypeId(uint32_t type_id) const
{
  if (!rtti_data_)
    return nullptr;

  uint8_t kind = type_id & kMaxTypeIdKind;
  uint32_t payload = (type_id >> 4) & kMaxTypeIdPayload;

  if (kind == kTypeId_Inline) {
    uint8_t bytes[4];
    bytes[0] = (payload >> 0) & 0xff;
    bytes[1] = (payload >> 8) & 0xff;
    bytes[2] = (payload >> 16) & 0xff;
    bytes[3] = (payload >> 24) & 0xff;

    RttiParser parser(bytes, 4, 0);
    return parser.decodeNew();
  } else if (kind == kTypeId_Complex) {
    RttiParser parser(rtti_data_, rtti_data_size_, payload);
    return parser.decodeNew();
  }
  return nullptr;
}

const Rtti*
RttiData::functionTypeFromOffset(uint32_t offset) const
{
  if (!rtti_data_)
    return nullptr;

  RttiParser parser(rtti_data_, rtti_data_size_, offset);
  return parser.decodeFunction();
}

const Rtti*
RttiData::typesetTypeFromOffset(uint32_t offset) const
{
  if (!rtti_data_)
    return nullptr;

  RttiParser parser(rtti_data_, rtti_data_size_, offset);
  return parser.decodeTypeset();
}

const uint8_t*
RttiData::blob() const
{
  return rtti_data_;
}

size_t
RttiData::size() const
{
  return rtti_data_size_;
}

bool
RttiData::validateType(uint32_t type_id) const
{
  uint8_t kind = type_id & kMaxTypeIdKind;
  uint32_t payload = (type_id >> 4) & kMaxTypeIdPayload;
  if (kind == kTypeId_Inline) {
    uint8_t bytes[4];
    bytes[0] = (payload >> 0) & 0xff;
    bytes[1] = (payload >> 8) & 0xff;
    bytes[2] = (payload >> 16) & 0xff;
    bytes[3] = (payload >> 24) & 0xff;

    RttiParser parser(bytes, 4, 0);
    return parser.validate();
  }
  else if (kind == kTypeId_Complex) {
    if (payload >= rtti_data_size_)
      return false;

    RttiParser parser(rtti_data_, rtti_data_size_, payload);
    return parser.validate();
  }
  else
    return false;
}

bool
RttiData::validateFunctionOffset(uint32_t offset) const
{
  RttiParser parser(rtti_data_, rtti_data_size_, offset);
  return parser.validateFunction();
}

bool
RttiData::validateTypesetOffset(uint32_t offset) const
{
  RttiParser parser(rtti_data_, rtti_data_size_, offset);
  return parser.validateTypeset();
}

RttiParser::RttiParser(const uint8_t* bytes, uint32_t length, uint32_t offset)
 : bytes_(bytes),
   length_(length),
   offset_(offset),
   is_const_(false)
{
}

// Decode a type, but reset the |is_const| indicator for non-
// dependent type.
Rtti*
RttiParser::decodeNew()
{
  bool was_const = is_const_;
  is_const_ = false;
  
  Rtti* result = decode();
  if (is_const_)
    result->setConst();

  is_const_ = was_const;
  return result;
}

Rtti*
RttiParser::decode()
{
  is_const_ = match(cb::kConst) || is_const_;

  uint8_t type = bytes_[offset_++];
  switch (type) {
  case cb::kBool:
  case cb::kInt32:
  case cb::kFloat32:
  case cb::kChar8:
  case cb::kAny:
  case cb::kTopFunction:
    return new Rtti(type);

  case cb::kFixedArray:
  {
    uint32_t size = decodeUint32();
    Rtti* inner = decode();
    return new Rtti(type, size, inner);
  }
  case cb::kArray:
  {
    Rtti* inner = decode();
    return new Rtti(type, inner);
  }
  case cb::kEnum:
  case cb::kTypedef:
  case cb::kTypeset:
  case cb::kClassdef:
  case cb::kEnumStruct:
  {
    uint32_t index = decodeUint32();
    return new Rtti(type, index);
  }
  case cb::kFunction:
    return decodeFunction();
  }
  return nullptr;
}

Rtti*
RttiParser::decodeFunction()
{
  uint8_t argc = bytes_[offset_++];

  bool variadic = false;
  if (bytes_[offset_] == cb::kVariadic) {
    variadic = true;
    offset_++;
  }

  Rtti* returnType;
  if (bytes_[offset_] == cb::kVoid) {
    returnType = new Rtti(cb::kVoid);
    offset_++;
  } else {
    returnType = decodeNew();
  }

  Rtti* functionType = new Rtti(returnType, variadic);
  for (uint8_t i = 0; i < argc; i++) {
    bool is_by_ref = match(cb::kByRef);
    Rtti* arg_type = decodeNew();
    if (is_by_ref)
      arg_type->setByRef();
    functionType->addArgument(arg_type);
  }
  return functionType;
}

Rtti*
RttiParser::decodeTypeset()
{
  uint32_t count = decodeUint32();
  Rtti* typeset = new Rtti(cb::kTypeset);

  for (uint32_t i = 0; i < count; i++) {
    typeset->addSignature(decodeNew());
  }
  return typeset;
}

bool
RttiParser::validate()
{
  if (offset_ >= length_)
    return false;
  // A type can start with a |const| indicator.
  match(cb::kConst);
  if (offset_ >= length_)
    return false;
  uint8_t type = bytes_[offset_++];
  switch (type) {
  case cb::kBool:
  case cb::kInt32:
  case cb::kFloat32:
  case cb::kChar8:
  case cb::kAny:
  case cb::kTopFunction:
    return true;

  case cb::kFixedArray:
  {
    // Skip the size.
    if (!tryDecodeUint32())
      return false;
    return validate();
  }
  case cb::kArray:
    return validate();

  case cb::kEnum:
  case cb::kTypedef:
  case cb::kTypeset:
  case cb::kClassdef:
  case cb::kEnumStruct:
    // Skip the index.
    return tryDecodeUint32();

  case cb::kFunction:
    return validateFunction();
  }
  return false;
}

bool
RttiParser::validateFunction()
{
  // argc available?
  if (offset_ >= length_)
    return false;

  uint8_t argc = bytes_[offset_++];
  if (offset_ >= length_)
    return false;

  if (bytes_[offset_] == cb::kVariadic)
    offset_++;
  if (offset_ >= length_)
    return false;

  // Validate return type.
  if (bytes_[offset_] == cb::kVoid)
    offset_++;
  else if (!validate())
    return false;

  for (uint8_t i = 0; i < argc; i++) {
    if (offset_ >= length_)
      return false;
    // A by_ref indicator is allowed here.
    match(cb::kByRef);
    if (!validate())
      return false;
  }
  return true;
}

bool
RttiParser::validateTypeset()
{
  // See if we can decode the count of signatures in this typset.
  uint32_t old_offset = offset_;
  if (!tryDecodeUint32())
    return false;

  // Decode the count now.
  offset_ = old_offset;
  uint32_t count = decodeUint32();
  for (uint32_t i = 0; i < count; i++) {
    if (!validate())
      return false;
  }
  return true;
}

bool
RttiParser::match(uint8_t b)
{
  if (bytes_[offset_] != b)
    return false;

  offset_++;
  return true;
}

uint32_t
RttiParser::decodeUint32()
{
  uint32_t value = 0;
  uint32_t shift = 0;
  while (true) {
    uint8_t b = bytes_[offset_++];
    value |= (b & 0x7f) << shift;
    if ((b & 0x80) == 0)
      break;
    shift += 7;
  }
  return value;
}

bool
RttiParser::tryDecodeUint32()
{
  while (offset_ < length_) {
    uint8_t b = bytes_[offset_++];
    if ((b & 0x80) == 0)
      return true;
  }
  return false;
}

Rtti::Rtti(uint8_t type)
 : type_(type),
   index_(0),
   inner_(nullptr),
   is_const_(false),
   is_by_ref_(false),
   is_variadic_(false)
{
}

Rtti::Rtti(uint8_t type, uint32_t size)
 : type_(type),
   index_(size),
   inner_(nullptr),
   is_const_(false),
   is_by_ref_(false),
   is_variadic_(false)
{
}

Rtti::Rtti(uint8_t type, Rtti* inner)
 : type_(type),
   index_(0),
   inner_(inner),
   is_const_(false),
   is_by_ref_(false),
   is_variadic_(false)
{
}

Rtti::Rtti(uint8_t type, uint32_t size, Rtti* inner)
 : type_(type),
   index_(size),
   inner_(inner),
   is_const_(false),
   is_by_ref_(false),
   is_variadic_(false)
{
}

Rtti::Rtti(Rtti* return_type, bool is_variadic)
 : type_(cb::kFunction),
   index_(0),
   inner_(return_type),
   is_const_(false),
   is_by_ref_(false),
   is_variadic_(is_variadic)
{
}

void
Rtti::setConst()
{
  is_const_ = true;
}

void
Rtti::setByRef()
{
  is_by_ref_ = true;
}

void
Rtti::addArgument(Rtti* arg)
{
  assert(type_ == cb::kFunction);
  args_.push_back(std::unique_ptr<const Rtti>(arg));
}

void
Rtti::addSignature(Rtti* arg)
{
  assert(type_ == cb::kTypeset);
  args_.push_back(std::unique_ptr<const Rtti>(arg));
}
